consensus/dbft/dbft.go

```
@@ -710,29 +711,32 @@ func (c *DBFT) verifyHeader(chain consensus.ChainHeaderReader, header *types.Hea
 	if header.Extra[0] != dbftutil.ExtraV0 {
 		return fmt.Errorf("unknown Extra version: %d", header.Extra[0])
 	}
-	m := crypto.GetBFTHonestNodeCount(len(c.config.StandByValidators))
-	sigBytesLen := m * extraSeal
-	if len(header.Extra) < dbftutil.ExtraVersionLen+sigBytesLen {
-		return errMissingSignature
-	}
-	// Ensure that the extra-data contains validators list.
-	signersBytes := len(header.Extra) - dbftutil.ExtraVersionLen - sigBytesLen
-	if signersBytes == 0 {
-		return fmt.Errorf("missing validators addresses")
-	}
-	if signersBytes%common.AddressLength != 0 {
-		return errInvalidExtraSigners
-	}
-	// Ensure that the mix digest is not zero.
-	if header.MixDigest == (common.Hash{}) {
-		return errInvalidMixDigest
+	if isSealed {
+		m := crypto.GetBFTHonestNodeCount(len(c.config.StandByValidators))
+		sigBytesLen := m * extraSeal
+		if len(header.Extra) < dbftutil.ExtraVersionLen+sigBytesLen {
+			return errMissingSignature
+		}
+		// Ensure that the extra-data contains validators list.
+		signersBytes := len(header.Extra) - dbftutil.ExtraVersionLen - sigBytesLen
+		if signersBytes == 0 {
+			return fmt.Errorf("missing validators addresses")
+		}
+		if signersBytes%common.AddressLength != 0 {
+			return errInvalidExtraSigners
+		}
+
+		// Ensure that the mix digest is not zero.
+		if header.MixDigest == (common.Hash{}) {
```

Can we have it in proposal and check local one against proposed?

---

We can, but why do we need this? MixDigest depends on validators list, nodes take validators list from the dBFT context, and if something is wrong with the validators list, the consensus just won't work. And nodes fill this field by their own before commit is send to include this field in the hashable part, exactly like in NeoGo. If malicious node wants to change this field, then Commit verification will fail on a good node. If malicios node will do everything right and intentionally change MixDigest after Commit release, then the resulting block relayed to the network by this node won't be accepted by other nodes because it has wrong MixDigest. But blocks relayed by other honest consensus nodes will be accepted.

So I don't think we need to include this field in PrepareRequest. And its stub will be removed from PrepareRequest in #82. Am I missing something here?

---

Fail fast. And better with some meaningful reason. Right that would be "wrong commit signature", go figure.

---

Should we perform the same update in NeoGo?

---

> Should we perform the same update in NeoGo?

OK, silly question, can't do this in NeoGo since NextConsensus is not a part of PrepareRequest and protocol extension is needed there. OK, will do it in Geth at least.

---

Fixed in a separate commit. But we need to recheck it once Governance contract is integrated and state-dependent validators calculations are available.

---

See https://github.com/bane-labs/go-ethereum/issues/14#issuecomment-1923566099.

---

It's going to be tested wrt validator list change anyway, I think.